src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -733,34 +741,32 @@ public async Task RunPacketReadLoopAsync()
             /// <param name="packet">The packet to send.</param>
             public void SendData(INodePacket packet)
             {
-                MemoryStream writeStream = new MemoryStream();
-                ITranslator writeTranslator = BinaryTranslator.GetWriteTranslator(writeStream);
+                // clear the buffer but keep the underlying capacity to avoid reallocations
+                _writeBufferMemoryStream.SetLength(0);
+
+                ITranslator writeTranslator = BinaryTranslator.GetWriteTranslator(_writeBufferMemoryStream);
                 try
                 {
-                    writeStream.WriteByte((byte)packet.Type);
+                    _writeBufferMemoryStream.WriteByte((byte)packet.Type);
 
                     // Pad for the packet length
-                    writeStream.Write(BitConverter.GetBytes((int)0), 0, 4);
+                    _writeBufferStreamWriter.Write(0);
                     packet.Translate(writeTranslator);
 
+                    int writeStreamLength = (int)_writeBufferMemoryStream.Position;
+
                     // Now plug in the real packet length
-                    writeStream.Position = 1;
-                    writeStream.Write(BitConverter.GetBytes((int)writeStream.Length - 5), 0, 4);
+                    _writeBufferMemoryStream.Position = 1;
+                    _writeBufferStreamWriter.Write(writeStreamLength - 5);
 
-                    byte[] writeStreamBuffer = writeStream.GetBuffer();
+                    byte[] writeStreamBuffer = _writeBufferMemoryStream.GetBuffer();
 
-                    for (int i = 0; i < writeStream.Length; i += MaxPacketWriteSize)
+                    for (int i = 0; i < writeStreamLength; i += MaxPacketWriteSize)
                     {
-                        int lengthToWrite = Math.Min((int)writeStream.Length - i, MaxPacketWriteSize);
-                        if ((int)writeStream.Length - i <= MaxPacketWriteSize)
+                        int lengthToWrite = Math.Min(writeStreamLength - i, MaxPacketWriteSize);
+                        if (writeStreamLength - i <= MaxPacketWriteSize)
```

Comment:
This `if` looks redundant now, both branches do the same thing.

Reply:
good point